Support for PKCS11

portersrc opened the pull request:
Initial pkcs11 support for ocicrypt-rs, ported from ocicrypt's golang version.

arronwy committed:
Add config module

Signed-off-by: Arron Wang <arron.wang@intel.com>

portersrc committed:
basic output/structure from cargo-new --lib

Signed-off-by: Chris Porter <porter@ibm.com>

arronwy committed:
Add helper module

Signed-off-by: Arron Wang <arron.wang@intel.com>

portersrc committed:
ignore swp

Signed-off-by: Chris Porter <porter@ibm.com>

portersrc committed:
merge with arron PR

Signed-off-by: Chris Porter <porter@ibm.com>

portersrc committed:
sync with arron's PR

Signed-off-by: Chris Porter <porter@ibm.com>

sameo committed:
Merge pull request #2 from arronwy/devel

Initial implementation for module spec/config/helper

Signed-off-by: R. Jantz <cporterbox@gmail.com>

sameo committed:
Merge pull request #2 from arronwy/devel

Initial implementation for module spec/config/helper

Signed-off-by: Chris Porter <porter@ibm.com>

portersrc committed:
Merge branch 'main' into pkcs11

Signed-off-by: Chris Porter <porter@ibm.com>

portersrc committed:
add dependencies for cargo

Signed-off-by: Chris Porter <porter@ibm.com>

portersrc committed:
add keywrap and utils modules to top-level lib.rs

Signed-off-by: Chris Porter <porter@ibm.com>

portersrc committed:
first cut of keywrapper trait

Signed-off-by: Chris Porter <porter@ibm.com>

portersrc committed:
first cut of pkcs11 module

Signed-off-by: Chris Porter <porter@ibm.com>

portersrc committed:
first cut of some helper functions for pkcs11; incompelte; needs reorganization

Signed-off-by: Chris Porter <porter@ibm.com>

portersrc committed:
add set-env-vars function

Signed-off-by: Chris Porter <porter@ibm.com>

fitzthum commented on src/utils.rs:
```diff
@@ -0,0 +1,128 @@
+//  - uri : <pkcs11 uri>                                                                  // An error is returned if the pkcs11 URI is malformed
+pub fn parse_pkcs11_key_file(yaml_bytes: &Vec<u8>)
+                             -> Result<Pkcs11KeyFileObject, std::io::Error> {
+    let s = serde_yaml::to_string(yaml_bytes).unwrap();
```
Is it safe to blindly unwrap this? We will get a panic at runtime if the yaml parsing doesn't work out. Might be good to pass the error upwards with the ? syntax. That may require changing the result type, but it doesn't look like you're using the error in the result anyway.

rudyjantz replied:
Thanks for the guidance on error handling. I've handled it in newer commits. I'm going to talk with Arron, too, to make sure we're on the same page about error handing.

fitzthum commented on src/utils.rs:
```diff
@@ -0,0 +1,128 @@
+    let kfo = Pkcs11KeyFileObject {
+        uri: p11_uri,
+    };
+    Ok(kfo)
```
You can probably do this all in one line.

fitzthum commented on src/keywrap/pkcs11/mod.rs:
```diff
@@ -0,0 +1,150 @@
+        let pkcs11_recipients: Vec<Pkcs11KeyFileObject>
+          = add_pub_keys(&(ec.decrypt_config.as_ref().unwrap()), &x).unwrap();
+
+        if pkcs11_recipients.len() == 0 {
```
vectors have an is_empty method that some people prefer over this kind of check.

fitzthum commented on src/keywrap/pkcs11/mod.rs:
```diff
@@ -0,0 +1,150 @@
+
+    Ok(pkcs11_keys)
+}
+
```
ik this function is not done yet, but it seems like maybe a good example of a very c-like approach. rather than individually adding things to a mutable vector, you might be able to return a match statement with `pub_keys.is_empty()` and either an empty vector or some map.

rudyjantz replied:
I updated this based on the suggestion. I also handled the similar problem in unwrap_keys. The version I came up with is harder for me to read. That could be because I'm not used to Rust or because my implementation is still bad.

fitzthum commented on src/keywrap/pkcs11/mod.rs:
```diff
@@ -0,0 +1,150 @@
+                      -> Result<Vec<String>, std::io::Error> {
+        let mut v = Vec::new();
+        v.push("[pkcs11]".to_string());
+        Ok(v)
```
you can probably turn these 3 lines into 1

fitzthum commented on src/keywrap/mod.rs:
```diff
@@ -0,0 +1,36 @@
+trait KeyWrapper {
+    fn wrap_keys(&self,
+                 ec: &EncryptConfig,
+                 opts_data: &Vec<u8>)
```
Is there any particular reason for using byte strings here?

arronwy committed:
Add keywrap trait

Signed-off-by: Arron Wang <arron.wang@intel.com>

arronwy committed:
Add keywrap trait

Signed-off-by: Arron Wang <arron.wang@intel.com>

arronwy committed:
keywrap: Add jwe module

Signed-off-by: Arron Wang <arron.wang@intel.com>

arronwy committed:
keywrap: Add jwe module

Signed-off-by: Arron Wang <arron.wang@intel.com>

arronwy committed:
cargo: Make jwe module configurable

Signed-off-by: Arron Wang <arron.wang@intel.com>

arronwy committed:
cargo: Make jwe module configurable

Signed-off-by: Arron Wang <arron.wang@intel.com>

arronwy committed:
data: Add more public/private keys for unit tests

Signed-off-by: Arron Wang <arron.wang@intel.com>

arronwy committed:
data: Add more public/private keys for unit tests

Signed-off-by: Arron Wang <arron.wang@intel.com>

portersrc committed:
crude cut for most or all of wrap-key support; untested, etc.

Signed-off-by: Chris Porter <porter@ibm.com>

portersrc committed:
resolve (at least temporarily) several warnings

Signed-off-by: rudy jantz <cporterbox@gmail.com>

arronwy committed:
Add github actions for PR check

Signed-off-by: Arron Wang <arron.wang@intel.com>

arronwy committed:
Add github actions for PR check

Signed-off-by: Arron Wang <arron.wang@intel.com>

arronwy committed:
Fix CI clippy warnings

Signed-off-by: Arron Wang <arron.wang@intel.com>

arronwy committed:
Fix CI clippy warnings

Signed-off-by: Arron Wang <arron.wang@intel.com>

portersrc committed:
replace unwrap calls with custom error; needs work

Signed-off-by: rudy jantz <cporterbox@gmail.com>

portersrc committed:
OcicrsError to OrsError for now

Signed-off-by: rudy jantz <cporterbox@gmail.com>

portersrc committed:
move ors-error to separate file

Signed-off-by: rudy jantz <cporterbox@gmail.com>

portersrc committed:
replace unwrap with custom error in keywrapper code

Signed-off-by: rudy jantz <cporterbox@gmail.com>

lumjjb commented on src/keywrap/pkcs11/mod.rs:
```diff
@@ -0,0 +1,152 @@
+        for p in ps {
+            x.push(p.to_vec());
+        }
+        for y in &ec.param["pkcs11-yamsl"] {
```
nit: typo?

rudyjantz replied:
good catch; gotta get the tests done...

lumjjb commented on Cargo.toml:
```diff
@@ -10,3 +10,11 @@ edition = "2018"
 [dependencies]
 anyhow = ">=1.0"
+pkcs11 = ">=0.5.0"
+pkcs11-uri = ">=0.1.2"
```
@stefanberger are these libraries using the same standard for PKCS11 URIs? This library says it's implementing RFC 7512, so it seems like there shouldn't be any issue here, but deferring to your opinion on the intricacies

stefanberger replied:
Following this page [here](https://nickray.github.io/pkcs11-uri/pkcs11_uri/index.html) it does:

```
PKCS#11 URI

Bare bones implementation of the RFC 7512 URI scheme for locating keys and other PKCS#11 objects.

This library is patched together from existing libraries, namely pkcs11, uriparse and percent-encoding, and is a work in progress.
```

So this library seems to want to do the same as mine does. Differences may be in the details.

lumjjb reviewed (commented):
At some point when this is ready to merge, it may be good for @stefanberger to take a quick look through as well as he was the main author of the PKCS11 component.

stefanberger commented:
> At some point when this is ready to merge, it may be good for @stefanberger to take a quick look through as well as he was the main author of the PKCS11 component.

My suggestion would be to look at the ocicrypt golang code and implement it just like it's implemented there, basically more or less a one-to-one. There were some slight differences in the behavior of pkcs11 modules that ended up driving the implementation to what it is. I have only used 2 (I think) different pkcs11 modules back then and the golang code ended up being able to use both of them but only after tweaking the code to adapt to the slight differences in behavior of these modules. The one module was an IBM-internal one (needed special access to their device) and the other one softhsm and maybe that I tried with the TPM 2 pkcs11 module, but I am not sure.

portersrc committed:
updates for unwrap code

Signed-off-by: rudy jantz <cporterbox@gmail.com>

portersrc committed:
patch up borrowing

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
simplify return statement

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
use is-empty instead of len == 0

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
create 1-element vec in 1 line

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
missed case of len == 0 conversion to is-empty

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
fix typo

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
quiet warnings for now

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
remove "get" from getter function names

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
Merge branch 'main' into pkcs11

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
change Pkcs11 struct to Pkcs11KeyWrapper to match jwe

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
add test for no-possible-keys

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
test for recipients; may not be necessary

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
unit test for annotation_id

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
tests for private keys and key ids from packet

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
reorder test

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
merge with main; broken commit

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
minimal changes to compile last merge

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
remove custom error idea for now

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
adding tempdir, rsa, rand, sha2 crates

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
add softhsm module; probably needs subdirectory

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
add pkcs11uriwrapped; this could potentially be a PR for the pkcs11-uri rust crate

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
pkcs11config and supporting funcs; may need to split out to a new file

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
round out some pkcs11 mod functionality, including (not yet passing) success test

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
remove custom error and add softhsm and pkcs11-uri-wrapped to top-level lib file

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
delete custom error (using anyhow in the project for now)

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
shoving pkcs11 functionality into utils. will need to be moved

Signed-off-by: rudy jantz <porter@gatech.edu>

dubek commented on src/pkcs11_uri_wrapped.rs:
```diff
@@ -0,0 +1,229 @@
+        }
+
+        let module_name = match &self.p11uri.query_attributes.module_name {
+            Some(mn) => mn.to_lowercase(),
```
not sure `to_lowercase` is a good approach here (and below) -- linux filesystem is case sensitive; you can have two files in a directory called "MODule_1" and "module_1".

dubek commented on src/pkcs11_uri_wrapped.rs:
```diff
@@ -0,0 +1,229 @@
+                    Ok(f) => f.to_lowercase(),
+                    Err(e) => continue,
+                };
+                let idx = match file_lower.find(&module_name) {
+                    Some(i) => i,
+                    None => continue,
+                };
+                // We require that file_lower ends with module_name or that
+                // a suffix follows so that softhsm will not match
+                // libsofthsm2.so but only libsofthsm.so
+                // FIXME: replace with idiomatic rust
+                if file_lower.len() == idx+module_name.len()
+                || file_lower.as_bytes()[idx+module_name.len()] == b'.' {
```
try to use std::path::Path methods here to check if a file is inside a dir or something like that.

dubek commented on src/softhsm.rs:
```diff
@@ -0,0 +1,141 @@
+const TEMPDIR_PREFIX: &str = "ocicrypt-rs";
+
+pub struct SoftHSMSetup {
+    pub statedir_path: String,
+    pub statedir_folder: TempDir,
```
maybe the String can be deduced from the TempDir (by adding a method `get_statedir_path(&self)` ? Why do we need both in the struct?

dubek commented on src/softhsm.rs:
```diff
@@ -0,0 +1,141 @@
+                let res = res[idx..].trim_end_matches(trim_me);
+                return Ok(res.to_string());
+            },
+            None => return Err(anyhow!("Failed to find 'pkcs11:' in output 
+                                       from `softhsm setup`")),
```
Need that `\` at the end of the first line otherwise the error string will have many spaces in the middle.

rudyjantz commented on src/config.rs:
```diff
@@ -280,3 +278,169 @@ mod tests {
+
+        match std::env::var(XDGCONFIGHOME) {
+            Ok(envvar) => {
+                let mut p = std::path::PathBuf::from(envvar);
```
needs to be mutable?

rudyjantz commented on src/config.rs:
```diff
@@ -280,3 +278,169 @@ mod tests {
+
+        match std::env::var(HOME) {
+            Ok(envvar) => {
+                let mut p = std::path::PathBuf::from(envvar);
```
try to make single line as well

rudyjantz commented on src/config.rs:
```diff
@@ -280,3 +278,169 @@ mod tests {
+        Err(anyhow!(""))
+    }
+
+    fn get_default_crypto_config_opts(&self) -> Result<Pkcs11Config> {
```
after testing passes, can replace with data/default.yaml kind-of-config

rudyjantz commented on src/config.rs:
```diff
@@ -280,3 +278,169 @@ mod tests {
+            module_directories: vec![],
+            allowed_module_paths: vec![],
+        };
+        let c = self.get_configuration()?;
```
look into the TODO, but also stick get-config in the OK() call.

rudyjantz commented on src/config.rs:
```diff
@@ -280,3 +278,169 @@ mod tests {
+            allowed_module_paths: vec![],
+        };
+        let c = self.get_configuration()?;
+        // TODO ? can we ever get an empty pkcs11config from get-configuration?
```
see parse_config_file()... the basically need to check if these 3rd party libraries can return an empty config. We need the "ic == nil" golang case, basically.

rudyjantz replied:
get-configuration could return an option, is one idea for this

rudyjantz replied:
Leaving as-is. The only place that GetUserPkcs11Config is invoked seems to be from the golang's parse-helpers `CreateCryptoConfig` and `CreateDecryptCryptoConfig`:
```CreateCryptoConfig
  pkcs11config.GetUserPkcs11Config
    getConfiguration
      parseConfigFile
```
This isn't used in the repo at the moment, it seems. And even there, it returns the error and an empty config. (So the entire chain of returns is passing along empty configs with an error.) We could just as easily handle the error in a "parse helper" if we ever make one. So there's no extra case to handle yet.

rudyjantz commented on src/config.rs:
```diff
@@ -280,3 +278,169 @@ mod tests {
+                              {} \
+                              allowed-module-paths:\n\
+                              {}", mdyaml, mdyaml);
+        parse_pkcs11_config_file(config.as_bytes())
```
may not even need this parse-pkcs11-config-file (which takes bytes) if we change this code (by just hard-coding a default yaml config for example).

rudyjantz commented on src/keywrap/pkcs11/mod.rs:
```diff
@@ -0,0 +1,333 @@
+        let mut pubkeys: Vec<Vec<u8>> = Vec::new();
+        match ec.param.get("pkcs11-pubkeys") {
+            Some(pks) => {
+                for p in pks {
```
consider how a map could replace this

rudyjantz commented on src/keywrap/pkcs11/mod.rs:
```diff
@@ -0,0 +1,333 @@
+            return Ok(Vec::new())
+        }
+
+        let json_str = encrypt_multiple(&pkcs11_recipients, opts_data)?;
```
just Ok. dont need json_str

rudyjantz commented on src/keywrap/pkcs11/mod.rs:
```diff
@@ -0,0 +1,333 @@
+
+        let mut pkcs11_keys = Vec::new();
+
+        let priv_keys = match self.private_keys(&dc.param) {
```
can this be done more neatly with a combinator?

rudyjantz commented on src/keywrap/pkcs11/mod.rs:
```diff
@@ -0,0 +1,333 @@
+        for key in priv_keys {
+
+            let mut k = parse_private_key(&key, &vec![], "PKCS11".to_string())?;
+            match k {
```
This match can also be an `if let`, I believe, which would look cleaner.
Also, we only use `k` when `p11conf_opt` has `Some` value. That check could be done outside this entire loop, presumably. But then that assumes `parse_private_key()` has no side effects, so check that. And finally, this begs the question Dov had below: Do we need to check that pkcs11_keys is non-empty?

rudyjantz commented on src/keywrap/pkcs11/mod.rs:
```diff
@@ -0,0 +1,333 @@
+        let shsm = vs.1;
+
+        // FIXME
+        //std::env::set_var("OCICRYPT_OAEP_HASHALG", "sha1");
```
if sha1 should be default, then where is that functionality coming from? which 3rd party library?

rudyjantz commented on src/keywrap/pkcs11/mod.rs:
```diff
@@ -0,0 +1,333 @@
+mod kw_tests {
+    use super::*;
+
+    //const SOFTHSM_SETUP: &str = "../../scripts/softhsm_setup";
```
Should we commit a copy of the softhsm_setup script? It currently exists in the ocicrypt golang repo.

rudyjantz commented on src/keywrap/pkcs11/mod.rs:
```diff
@@ -0,0 +1,333 @@
+    fn test_keywrap_pkcs11_success() {
+        let path_to_script = env!("CARGO_MANIFEST_DIR").to_string() + "/scripts/softhsm_setup";
+        let vs = create_valid_pkcs11_ccs().unwrap();
+        let valid_pkcs11_ccs = vs.0;
```
use unpack assignment

rudyjantz commented on src/keywrap/pkcs11/mod.rs:
```diff
@@ -0,0 +1,333 @@
+
+        }
+
+        shsm.run_softhsm_teardown(&path_to_script);
```
The paired `setup` call actually comes from `create_valid_pkcs11_ccs`, which is confusing (and also led to me having to create two `path_to_script` variables. Fix this.

rudyjantz commented on src/keywrap/pkcs11/mod.rs:
```diff
@@ -0,0 +1,333 @@
+    fn get_pkcs11_config_yaml() -> Result<Vec<u8>> {
+        // we need to provide a configuration file so that on the various
+        // distros the libsofthsm2.so will be found by searching directories
+        let mdyaml = get_default_module_directories_yaml("".to_string())?;
```
similar to previous comments: create the config object, and then use serde to serialize as needed

rudyjantz commented on src/pkcs11_uri_wrapped.rs:
```diff
@@ -0,0 +1,229 @@
+    // whether a PIN has been provided at all so that an error code returned by
+    // this function indicates that the PIN value could not be retrieved.
+    pub fn pin(&self) -> Result<String> {
+        match &self.p11uri.query_attributes.pin_value {
```
if let Some

rudyjantz commented on src/utils.rs:
```diff
@@ -0,0 +1,691 @@
+        return None;
+    }
+    let oldenv = std::env::vars();
+    for (key, value) in std::env::vars() {
```
bug: should be for key,val in env (not std-env-vars)

rudyjantz commented on src/utils.rs:
```diff
@@ -0,0 +1,691 @@
+                          None,
+                          None)?;
+    if !pin.is_empty() {
+        match p11ctx.login(session, pkcs11::types::CKU_USER, Some(&pin)) {
```
if let

rudyjantz commented on src/keywrap/pkcs11/mod.rs:
```diff
@@ -0,0 +1,333 @@
+        let valid_pkcs11_ccs = vs.0;
+        let shsm = vs.1;
+
+        // FIXME
```
bigger issue and question for the whole repo: why are we setting environment variables inside of ocicrypt?

rudyjantz commented on src/lib.rs:
```diff
@@ -1,4 +1,8 @@
+#[macro_use] extern crate serde_derive;
```
can this be more nicely done by putting in toml file perhaps, or can it at least be moved down?

rudyjantz commented on src/pkcs11_uri_wrapped.rs:
```diff
@@ -0,0 +1,229 @@
+                    Err(e) => continue,
+                };
+                let file_lower = match file.file_name().into_string() {
+                    Ok(f) => f.to_lowercase(),
```
dot chain this stuff

rudyjantz commented on src/pkcs11_uri_wrapped.rs:
```diff
@@ -0,0 +1,229 @@
+                    Err(e) => continue,
+                };
+                let idx = match file_lower.find(&module_name) {
+                    Some(i) => i,
```
combinator with unwrap_or maybe?

portersrc committed:
add sha-1 crate

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
fix to sha1 in test case

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
basic success test case passes; 3 major fixes: env bug fix, switch to sha1 (256 not supported), and proper mechanism args

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
add softhsm_setup script (copy of the version in golang ocicrypt repo)

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
rustfmt

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
add missing colon

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
pull errors out of unnecessary nest

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
better pathbuf creation

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
add missing func call

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
remove temp variable and reduce return statement to one line

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
improve assignment syntax

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
simplify match expression

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
simplify match blocks for wrap_keys

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
remove unneeded temp variables before returns

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
replace match with ok_or_else

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
replace for loop with map in add-pub-keys

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
bug fix: wrap-keys pushes pkcs11 keys as long as KeyType is pkfo

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
remove unneeded assignment

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
change for loop to map in unwrap keys

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
use SOFTHSM_SETUP constant

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
tuple assignment

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
return Self from constructor

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
use clone instead of to_vec

Signed-off-by: rudy jantz <porter@gatech.edu>

arronwy committed:
Add blockcipher module

Signed-off-by: Arron Wang <arron.wang@intel.com>

arronwy committed:
Add blockcipher module

Signed-off-by: Arron Wang <arron.wang@intel.com>

arronwy committed:
blockcipher: Add aes ctr mode support

Signed-off-by: Arron Wang <arron.wang@intel.com>

arronwy committed:
blockcipher: Add aes ctr mode support

Signed-off-by: Arron Wang <arron.wang@intel.com>

arronwy committed:
Fix clippy warnings for test code

Signed-off-by: Arron Wang <arron.wang@intel.com>

arronwy committed:
Fix clippy warnings for test code

Signed-off-by: Arron Wang <arron.wang@intel.com>

portersrc committed:
fix comment and make type explicit

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
improve pin()

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
improve module() (syntax, dont use to-lowercase)

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
fmt

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
improve syntax for pkcs11-open-session

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
dont cache statedir-folder string

Signed-off-by: rudy jantz <porter@gatech.edu>

rudyjantz commented on src/keywrap/pkcs11/mod.rs:
```diff
@@ -0,0 +1,333 @@
+    }
+
+
+    fn annotation_id(&self) -> &str {
+        "org.opencontainers.image.enc.keys.pkcs11"
+    }
```
I think this guy has to be here for the `impl` to be complete. When we integrate with the actual users then I think you're right -- then we might have to set `pub` for whatever gets exported.

portersrc committed:
add license info

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
improve is-allowed-path

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
add error msg

Signed-off-by: rudy jantz <porter@gatech.edu>

arronwy committed:
Add base64 encode/decode for Vec<u8> during serializing

golang marshal []byte to JSON format will auto base64 encoded:
https://pkg.go.dev/encoding/json#Marshal

Signed-off-by: Arron Wang <arron.wang@intel.com>

arronwy committed:
Add base64 encode/decode for Vec<u8> during serializing

golang marshal []byte to JSON format will auto base64 encoded:
https://pkg.go.dev/encoding/json#Marshal

Signed-off-by: Arron Wang <arron.wang@intel.com>

sameo committed:
gh: Create dco.yml

Signed-off-by: Samuel Ortiz <samuel.e.ortiz@protonmail.com>

sameo committed:
gh: Create dco.yml

Signed-off-by: Samuel Ortiz <samuel.e.ortiz@protonmail.com>

portersrc committed:
trivial improvements; fix all warnings

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
disable softhsm-related tests; checking github tests pass

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
merge with main

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
satisfy all but 1 clippy error

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
cargo fmt

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
fix clippy box warning; disable remaining pkcs11 tests to check github pass status

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
cargo fmt

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
Merge branch 'main' into pkcs11

Signed-off-by: R. Jantz <cporterbox@gmail.com>

portersrc committed:
merge after force

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
merge with main

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
drop macro-use

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
drop extra serde-derive

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
fix build issues after merge

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
delete unnecessary comment

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
full commit of pkcs11 doc from golang version

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
cargo fmt

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
change ocicrypt to ocicrypt-rs in the doc

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
improved comments for pkcs11 keywrap module

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
improve comments for config.rs

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
comment out broken env variable test

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
reinstate process-private-keyfiles code

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
improve comments for pkcs11_uri_wrapped

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
improve comments for softhsm.rs

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
improve comments for pkcs11 utils module

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
improve has-pin by using is-some

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
fix string spacing

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
use struct directly instead of creating yaml string and converting

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
remove redundant clone

Signed-off-by: rudy jantz <porter@gatech.edu>

portersrc committed:
use pkcs11config struct instead of yaml string in test code

Signed-off-by: rudy jantz <porter@gatech.edu>

rudyjantz commented:
@lumjjb @fitzthum this pkcs11 support never got merged in August, but I want to find out if it's still wanted/needed in ocicrypt-rs. If so, what's a good next set of steps? There's a lot of code here. I can confirm that the happy-path test replicated from golang's ocicrypt works.

Also, what is the slack URL for discussing ocicrypt development? I cannot seem to find it.

Lastly, testing environment is a problem in this PR, and I'm open to suggestions. I disabled a test that's unrelated to my PR but which was breaking the automatic github push checkers (caf91a4). I think this is due to an environment variable problem from here:
    https://github.com/containers/ocicrypt-rs/blob/main/src/encryption.rs#L26-L28
Similarly, tests for the pkcs11 support in this PR depend on softhsm and create temporary files in /tmp, for example. These non-unit tests break the github checkers, as well, so I've disabled them.

rudyjantz commented:
@arronwy @sameo  - Please feel free to comment on my previous message. But in particular, if you have a pointer to the slack URL for confidential containers and ocicrypt-rs, that would be very helpful. Thanks.

portersrc committed:
merge with main

portersrc committed:
remove duplicate in cargo.toml

Signed-off-by: chris porter <porter@gatech.edu>

portersrc committed:
remove more duplicates from cargo.toml

Signed-off-by: chris porter <porter@gatech.edu>

portersrc committed:
temporarily skip use sha1 instead of sha256 due to error; improve syntax

Signed-off-by: chris porter <porter@gatech.edu>

portersrc committed:
remove unnecessary tostring call

Signed-off-by: chris porter <porter@gatech.edu>

portersrc committed:
comment out no-possible-keys (no longer a trait?); add probe TODO (just return false for now)

Signed-off-by: chris porter <porter@gatech.edu>

portersrc committed:
fix rsa to 0.5.0

Signed-off-by: chris porter <porter@gatech.edu>

portersrc committed:
bump rsa version to 0.6.1

Signed-off-by: chris porter <porter@gatech.edu>

portersrc committed:
change rsa::pkcs8 import to match latest version; put the FIXME for sha1 now instead of sha256

Signed-off-by: chris porter <porter@gatech.edu>

portersrc committed:
format

Signed-off-by: chris porter <porter@gatech.edu>

portersrc committed:
fix rustc 1.62.0-beta.4 clippy errors

Signed-off-by: chris porter <porter@gatech.edu>

portersrc committed:
attempt to fix rustc 1.63.0-nightly clippy errors

Signed-off-by: chris porter <porter@gatech.edu>

portersrc committed:
add missing Eq

Signed-off-by: chris porter <porter@gatech.edu>

portersrc committed:
fix sha1 version

Signed-off-by: chris porter <porter@gatech.edu>

portersrc committed:
reinstate sha1 padding

Signed-off-by: chris porter <porter@gatech.edu>

portersrc commented:
The "nightly" github job may have a small bug which is causing it to fail. Here's its error:
```error: you are deriving `PartialEq` and can implement `Eq`
Error:  --> src/utils/keyprovider.rs:1:17
  |
1 | #[derive(Clone, PartialEq, ::prost::Message)]
  |                 ^^^^^^^^^ help: consider deriving `Eq` as well: `PartialEq, Eq````
But I've already made this change [here](https://github.com/confidential-containers/ocicrypt-rs/pull/3/commits/85a48515a8bd06116a9e68cedce3e6a7c184f11b). Note that clippy modifies the keyprovider.rs file, so what may be happening is a prior github job (i.e. stable or beta) modifies the source, which then gets (incorrectly) reused in the nightly job.

portersrc commented:
@jodh-intel
This PR is up-to-date again. Is there a lead developer on the repo now? Or: Is there a customer/user for pkcs11 that you know of?
